Use scikit-build

pablogsal opened the pull request:
- **Try to respect RPATHS of calling dlopen modules with dlinfo**
- **Use scikit-build-core**

*Issue number of the reported bug or feature request: #<number>*

**Describe your changes**
A clear and concise description of the changes you have made.

**Testing performed**
Describe the testing you have performed to ensure that the bug has been addressed, or that the new feature works as planned.

**Additional context**
Add any other context about your contribution here.

pablogsal committed:
Use scikit-build-core

Signed-off-by: Pablo Galindo <pablogsal@gmail.com>

godlygeek commented on src/memray/_memray.pyx:
```diff
@@ -69,12 +69,12 @@ from libcpp.unordered_map cimport unordered_map
-from ._metadata import Metadata
-from ._stats import Stats
-from ._thread_name_interceptor import ThreadNameInterceptor
+from memray._destination import Destination
+from memray._destination import FileDestination
+from memray._destination import SocketDestination
+from memray._metadata import Metadata
+from memray._stats import Stats
+from memray._thread_name_interceptor import ThreadNameInterceptor
```
Something about how `scikit-build-core` does editable installs is causing this to be necessary... Things work fine with a non-editable install without these changes, but not with an editable install.

godlygeek replied:
Ah, presumably this:

> The default `editable.mode`, "redirect", uses a custom redirecting finder to combine the static CMake install dir with the original source code.

It's probably winding up wrong about what its package is due to that...

godlygeek replied:
But somehow it's working for `memray._test_utils`, which _also_ does
```py
from ._destination import Destination
```
Hm, there's still something I don't understand here...

godlygeek replied:
Ah, the thing that's throwing it off is that there's both a `memray/_memray/` directory and a `memray._memray` module, and they're not the same. I think this is a bug in `scikit-build-core`'s editable mode. I'll try to whip up an MRE to report the bug to them when I can. It should be easy to demonstrate that `__package__` changes depending on whether or not the install is editable.

godlygeek replied:
That's been fixed in scikit-build-core 0.10.6

godlygeek committed:
Drop CYTHON_TEST_MACROS / MEMRAY_MINIMIZE_INLINING

These environment variables previously allowed us to conditionally
enable extra safety checks in our Cython code, or conditionally disable
optimizations that might break tools like Valgrind. Now, we control this
based on the CMake build type, so that release builds have the extra
optimizations and debug builds don't, and so that debug builds have the
extra safety checks and release builds don't.

Signed-off-by: Matt Wozniski <mwozniski@bloomberg.net>

godlygeek committed:
tests: Avoid clobbering environment variables

Some tests were erroneously removing all environment variables from the
environment and passing along an environment consisting of only a single
variable. Since this removes environment variables like $PATH, this
breaks automatic rebuilds -- they can't find the compiler!

Signed-off-by: Matt Wozniski <mwozniski@bloomberg.net>

godlygeek committed:
tests: Loosen a check on elapsed time

This test failed in CI because the same time was captured in both
`end_time` and `start_time`. There's no reason to assume that this took
more than a millisecond, or that the system clock can give us better
than millisecond granularity, so it seems more reasonable to simply
check that the end time is no earlier than the start time. Technically
even that assumption could be violated, since we're not using
a monotonic clock, but oh well, this is just tests.

Signed-off-by: Matt Wozniski <mwozniski@bloomberg.net>

godlygeek committed:
Remove our Makefile from .gitignore

We meant to ignore any Makefile generated by CMake, but we shouldn't be
ignoring our own hand-written Makefile.

Signed-off-by: Matt Wozniski <mwozniski@bloomberg.net>

godlygeek committed:
docs: Drop the memray3.N entry point

With `scikit-build-core`, we don't have any way to produce an entry
point script called `memray3.N` (where `3.N` is the Python version
Memray was installed for). Drop references to it from the docs.

Signed-off-by: Matt Wozniski <mwozniski@bloomberg.net>

godlygeek committed:
Find `_inject.abi3.so` relative to `_memray.*.so`

With an editable `scikit-build-core` install, shared libraries are not
in the same directory as Python files, so we can no longer locate it
relative to `__init__.py`.

Signed-off-by: Matt Wozniski <mwozniski@bloomberg.net>

godlygeek commented:
This is missing the `npm install` and `npm run-script build` steps, I think

pablogsal commented:
I rebased  and fixes all the other issues

pablogsal commented:
@godlygeek this is ready can you approve?

godlygeek commented on src/memray/commands/_attach.gdb:
```diff
@@ -19,7 +19,7 @@ set scheduler-locking on
 
 p "Checking if we can access the library"
-set $result = (int)access($libpath, 5)
+set $result = (int)access($libpath, 4)
```
I don't understand this change. Is `scikit-build-core` sometimes creating shared libraries that don't have the executable bit set? If so, I can't reproduce that... What circumstances does it do that under?

pablogsal replied:
You are right: this is not a `scikit-build-core` behavior. CMake, the wheel, and a clean UV 3.14 editable install all keep the execute bit. The rebased UV and coverage jobs failed at this guard, but their logs did not record the file mode. The actual issue is that `5` checks `R_OK | X_OK`, while the guard only needs readability; I verified `dlopen` succeeds on a `0644` library. This is an unrelated forward-port fix.

godlygeek replied:
Ah, I see what caused this. https://github.com/bloomberg/memray/pull/959 fixes it. Though I've left your change in place, since you're right, we only need `R_OK` and not `R_OK | X_OK`, but I've figured out why the `x` bit wasn't (always?) getting set, and it's an easy fix.

godlygeek commented on pyproject.toml:
```diff
@@ -1,19 +1,130 @@
+  "pytest",
+  "pytest-cov",
+  "ipython",
+  "setuptools",
```
```suggestion
  "scikit-build-core",
```

godlygeek commented on pyproject.toml:
```diff
@@ -1,19 +1,130 @@
+  "pytest",
+  "pytest-cov",
+  "ipython",
+  "setuptools",
```
```suggestion
  "scikit-build-core",
```

godlygeek commented on pyproject.toml:
```diff
@@ -1,19 +1,130 @@
+wheel.exclude = ["memray/_memray/**"]
 
 [dependency-groups]
 test=['Cython', 'coverage[toml]', "greenlet; python_version < '3.15'", 'pytest', 'pytest-cov', 'ipython', 'setuptools', 'pkgconfig', 'pytest-textual-snapshot', 'textual >= 0.43, != 0.65.2, != 0.66', 'packaging']
```
```suggestion
test=['Cython', 'coverage[toml]', "greenlet; python_version < '3.15'", 'pytest', 'pytest-cov', 'ipython', 'scikit-build-core', 'pkgconfig', 'pytest-textual-snapshot', 'textual >= 0.43, != 0.65.2, != 0.66', 'packaging']
```

godlygeek reviewed (changes requested):
Just a few small things:

The `setup.py` used to run `npm install` and `npm run-script build` if `src/memray/reporters/templates/assets/*.js` didn't exist. The CMakeLists.txt never does that. Is that an intentional omission? Is the assumption that, because these files are checked into source control, we should never encounter the situation where they're missing, and so what was in the `setup.py` was dead code?

And, on a separate note: if we're removing the `MANIFEST.in`, we should also remove the `setup.py`, shouldn't we? Is there a reason to keep it?

pablogsal commented:
Done. I removed `setup.py`, updated ASV, and verified all checked-in JS assets in the sdist/wheel and the regeneration stuff. I added `scikit-build-core` but kept `setuptools` because the integration fixtures still use `setup.py`/`distutils`.

godlygeek commented:
> kept `setuptools` because the integration fixtures still use `setup.py`/`distutils`

Ah, right, makes sense

godlygeek reviewed (approved):
LGTM. I copied the big comment about our TLS model hackery over from the setup.py into the CMakeLists.txt, and squashed a few of these commits together. I left the ones that were fixing separate issues that were merely unearthed by the build backend change un-squashed.
